### PR TITLE
Gen

### PR DESCRIPTION
**Pull request by zainh96**

### CODE REVIEW

File: abstractrule.cpp [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+#include <iostream>
+
+using namespace rapidjson;
+using namespace std;
+
+namespace Catan {
+	namespace Generate {
+		AbstractRule::AbstractRule(map<string, bool> &boolMap) {
+			// Only need the "follow_strictly" field.
+			map<string, bool>::iterator it = boolMap.find("follow_strictly");
```
Comment (0aix) [inline abstractrule.cpp]: auto

File: boardgraph.cpp [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+			while (positionLeft > 0) {
+				leftCurrent = leftCurrent->neighbours[S_INDEX];
+				positionLeft--;
+			}
+
+			while (positionRight > 0) {
+				rightCurrent = rightCurrent->neighbours[S_INDEX];
+				positionRight--;
+			}
```
Comment (0aix) [inline boardgraph.cpp]: why are these not for loops?

File: boardgraph.cpp [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+				leftCurrent = leftCurrent->neighbours[S_INDEX];
+				rightCurrent = rightCurrent->neighbours[S_INDEX];
+			}
+		}
+
+		BoardGraphForwardIterator &BoardGraph::ForwardIterator() {
+			return *(new BoardGraphForwardIterator(firstNodes));
+		}
+
+    int RandomIterator(int i) {
```
Comment (0aix) [inline boardgraph.cpp]: make this a lambda

File: boardgraph.cpp [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+			}
+
+			for (int i = 0; i < config->chits.size(); i++) {
+				if (config->chits[i] > 0) {
+					chitList.insert(chitList.end(), config->chits[i], i + 1);
+				}
+			}
+
+			random_shuffle(typesList.begin(), typesList.end(), RandomIterator);
+			random_shuffle(chitList.begin(), chitList.end(), RandomIterator);
```
Comment (0aix) [inline boardgraph.cpp]: auto RandomIterator = [](int& i)
{
return rand() % i;
};

File: boardgraph.cpp [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+				if (node->CanPlaceChit()) {
+					node->chit = chitList[chitPos++];
+				}
+			}
+		}
+
+		int BoardGraph::BoardHeight() {
+			int height = 0;
+			for (int col : config->boardColumns) {
+				height = max(height, col);
```
Comment (0aix) [inline boardgraph.cpp]: tsk. assigns every time. wouldnt want to see something like that in calcite...

File: boardnode.cpp [hunk 6]
```diff
@@ -0,0 +1,37 @@
+#include "include/boardnode.h"
+#include <vector>
+
+using namespace std;
+
+namespace Catan {
+	namespace Generate {
+		BoardNode::BoardNode() {
+			for (int i = 0; i < NEIGHBOUR_COUNT; i++) {
```
Comment (0aix) [inline boardnode.cpp]: is that a U

File: include/boardnode.h [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+		//       N
+		//     _____
+		// NW /     \ NE
+		//   /       \
+		//   \       /
+		// SW \_____/ SE
+		//       S
+		class BoardNode {
+		public:
+			// Useful for graph traversal
```
Comment (0aix) [inline include/boardnode.h:36]: do your constructor work here

File: boardnode.cpp [hunk 8]
```diff
@@ -0,0 +1,37 @@
+#include "include/boardnode.h"
+#include <vector>
+
+using namespace std;
+
+namespace Catan {
+	namespace Generate {
+		BoardNode::BoardNode() {
```
Comment (0aix) [inline boardnode.cpp]: constructor might be unnecessary. could be done in header file

File: include/textrenderer.h [hunk 9]
```diff
@@ -0,0 +1,106 @@
+#pragma once
```
Comment (0aix) [inline include/textrenderer.h:1]: wtf header file?

File: islandcountrule.cpp [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+#include <map>
+#include <string>
+
+using namespace std;
+
+namespace Catan {
+	namespace Generate {
+		IslandCountRule::IslandCountRule(map<string, int> &intMap, map<string, bool> &boolMap) : AbstractRule(boolMap) {
+			// Populate min & max
+			map<string, int>::iterator it = intMap.find("min");
```
Comment (0aix) [inline islandcountrule.cpp]: auto

File: main.cpp [hunk 11]
```diff
@@ -1,14 +1,66 @@
...
+				cerr << "Usage: -r [text|image]" << endl;
+				return CLI_ARG_ERROR_CODE;
+			}
+		}
+	}
+
+	Catan::Generate::BoardGraph graph = Catan::Generate::BoardGraph(&config);
+	graph.Randomize();
+
+	if (renderType == "text") {
```
Comment (0aix) [inline main.cpp:58]: renderType.compare("text")

File: main.cpp [hunk 12]
```diff
@@ -1,14 +1,66 @@
...
+		Catan::Draw::RenderAsText(graph);
+	} else if (renderType == "image" && Catan::Draw::Initialize()) {
```
Comment (0aix) [inline main.cpp:60]: above

File: main.cpp [hunk 13]
```diff
@@ -1,14 +1,66 @@
...
+// -c [Config file path]
+// -r [text|image]
+int main(int argv, char **argc)
 {
-    if (Catan::Draw::Initialize())
-    {
+  srand(time(NULL));
+	Catan::Generate::Config config;
+	string renderType = "text";
+	// Handle command line arguments
```
Comment (0aix) [inline main.cpp:26]: if you didn't know, you can handle command line arguments in C and C++ w/ getopt

File: main.dSYM/Contents/Info.plist [hunk 14]
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (0aix) [inline main.dSYM/Contents/Info.plist]: remove..

File: nochitstouchingrule.cpp [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+#include "include/boardnode.h"
+#include <map>
+#include <string>
+#include <vector>
+
+using namespace std;
+
+namespace Catan {
+	namespace Generate {
+		NoSpecialChitsTouchingRule::NoSpecialChitsTouchingRule(map<string, bool> &boolMap) : AbstractRule(boolMap) {
```
Comment (0aix) [inline nochitstouchingrule.cpp]: you can just have AbstractRule(boolMap) in the header then

File: nochitstouchingrule.cpp [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+		}
+
+		bool NoSpecialChitsTouchingRule::IsFollowed(BoardGraph &graph) {
+			BoardGraphForwardIterator it = graph.ForwardIterator();
+			while (it.HasNext()) {
+				BoardNode *node = it.Next();
+				if (IsSpecialChit(node->chit)) {
+					// Check Neighbours
+					vector<BoardNode*> neighbours = node->NonNullNeighbours();
+					for (BoardNode *neighbour : neighbours) {
```
Comment (0aix) [inline nochitstouchingrule.cpp]: for (BoardNode* neighbour : node->NonNullNeighbours())

File: nopondsrule.cpp [hunk 17]
```diff
@@ -0,0 +1,18 @@
+#include "include/nopondsrule.h"
+#include "include/boardgraph.h"
+#include <map>
+#include <string>
+
+using namespace std;
+
+namespace Catan {
+	namespace Generate {
+		NoPondsRule::NoPondsRule(std::map<std::string, bool> &boolMap) : AbstractRule(boolMap) {
```
Comment (0aix) [inline nopondsrule.cpp]: same as above

File: boardgraph.cpp (hunk 5 above)
Comment (zainh96) [reply boardgraph.cpp]: Want me to just compute it once? k..

File: boardnode.cpp (hunk 6 above)
Comment (zainh96) [reply boardnode.cpp]: I'm Canadian

File: include/textrenderer.h (hunk 9 above)
Comment (zainh96) [reply include/textrenderer.h:1]: Yeah, didn't really make sense to put it in a class tbh

File: main.cpp (hunk 11 above)
Comment (zainh96) [reply main.cpp:58]: https://stackoverflow.com/questions/9158894/differences-between-c-string-and-compare

File: main.cpp (hunk 12 above)
Comment (zainh96) [reply main.cpp:60]: same

File: main.cpp (hunk 13 above)
Comment (zainh96) [reply main.cpp:26]: ok ill look into it

File: main.dSYM/Contents/Info.plist (hunk 14 above)
Comment (zainh96) [reply main.dSYM/Contents/Info.plist]: sry

File: boardgraph.cpp (hunk 2 above)
Comment (zainh96) [reply boardgraph.cpp]: idk

File: main.cpp (hunk 11 above)
Comment (0aix) [reply main.cpp:58]: huh i tried looking for it on the reference, but i didnt find it (the first time) ..

**Pull request merged**